# Perth ResBaz 2019 - Python Stream


This lesson is adapted from the [Data Carpentry Ecology lesson](http://www.datacarpentry.org/python-ecology-lesson/)

We'll be using an EtherPad channel to share solutions to challenges, ask questions and chat:

**enter link here**


## Download this notebook!

We'll be working from this notebook and filling in the answers as we go.
To fetch the notebook and the example data we will be using...

### Please download the Git repository at <https://github.com/darcyabjones/resbaz-perth-2019-python>.

If you have `git` installed you can `clone` it from the terminal at the location that you want to store it...

```bash
git clone https://github.com/darcyabjones/resbaz-perth-2019-python.git
```


If you don't have `git` installed, you can download a Zip archive of the repository from <https://github.com/darcyabjones/resbaz-perth-2019-python/archive/master.zip> and unzip it wherever you want :).



## How to use a Jupyter Notebook

https://jupyter-notebook-beginner-guide.readthedocs.io/en/latest/index.html

https://www.packtpub.com/books/content/getting-started-jupyter-notebook-part-1

- The file autosaves
- You run a cell with **Ctrl + enter** (or **Cmd + enter** on Mac OS) or using the run button in the tool bar
- If you run a cell with **Shift + enter** it will also create a new cell below
- See *Help > Keyboard Shortcuts* or the *Cheatsheet* for more info


- The notebook has different type of cells: Code and Markdown are most commonly used
- **Code** cells expect code for the Kernel you have chosen, syntax highlighting is available, comments in the code are specified with # -> code after this will not be executed
- **Markdown** cells allow you to right report style text, using markdown for formatting the style (e.g. Headers, bold face etc)

## Introduction to Python and data analysis using pandas

Python is a high-level, interpreted programming language. This means the code is easy to read for humans and there is no need for us to compile it and in many cases we do not have to think too much about the underlying system fro e.g. memory usage.

As a consequence, we can use it in two ways:
- Using the interpreter as an "advanced calculator" in interactive mode:

In [1]:
# Calculations
2 + 1

3

In [2]:
# Writing things to screen
print("Hello")

Hello


- Executing programs/scripts saved as a text file, usually with *.py extension:

In [5]:
%%bash
# running scripts
# Write a "Hello World" script using the Jupyter text editor.
# run it from a bash terminal using the %%bash cell magic.

cat <<EOF > hello_world.py
print("Hello")
EOF

python3 ./hello_world.py

Hello


# Types of Data

How information is stored in a DataFrame or a python object affects what we can do with it and the outputs of calculations as well. There are two main types of data that we're explore in this lesson: numeric and character types.


## Numeric Data Types

Numeric data types include integers and floats.
A **floating point** (known as a float) number has decimal points even if that decimal point value is `0`.
For example: `1.13`, `2.0`, `1234.345`.
If we have a column that contains both integers and floating point numbers, Pandas will assign the entire column to the float data type so the decimal points are not lost.
In a vector or data fram (we learn about these different types later) the entire object or an entire column will be of the same type.

An **integer** will never have a decimal point.
Thus `1.13` would be stored as `1`, and `1234.345` is stored as `1234`.
You will often see the data type `Int64` in python which stands for 64 bit integer.
The 64 simply refers to the amount of memory allocated to store data in each cell which effectively relates to how many digits it can store in each "cell".
Allocating space ahead of time allows computers to optimize storage and processing efficiency.


## Character Data Types

Strings are values that contain numbers and/or characters. 
For example, a string might be a word, a sentence, or several sentences. 
A string can also contain or consist of numbers.
For instance, `'1234'` could be stored as a string, as could `'10.23'`.
However, **strings that contain numbers can not be used for mathematical operations**!

In [11]:
# Assign an int, a float, and a str to the variables: number, pi_value, and text.

number = 2
pi_value = 3.14
text = "Data Carpentry"  # Note double quotes and single are the same in python. Just be consistent. 

Here we've assigned data to variables, namely `text`, `number` and `pi_value`, using the assignment operator `=`.
The variable called `text` is a string which means it can contain letters and numbers.
We could reassign the variable `text` to an integer too - but be careful reassigning variables as this can get confusing.

To print out the value stored in a variable we can simply type the name of the variable into the interpreter:

In [8]:
text

'Data Carpentry'

A cell, by default, will print to screen the last thing it evaluates (unless this is explicitly written to a variable).

Thus, in scripts and for evaluating things anywhere else within a cell, we must use the `print` function:

In [12]:
# Next line will print out text
print(text)

Data Carpentry


In [13]:
# We also need the print statement if we want to see more than one variable
text
number

2

In [16]:
print(text)
print(number)

Data Carpentry
2


In [14]:
print(text, number, pi_value)

Data Carpentry 2 3.14


### Operators

We can perform mathematical calculations in Python using the basic operators
 `+, -, /, //, *, %`:

In [18]:
6 * 7  # multiplication
2**16  # exponents
13 % 5  # Modulo

3

In [24]:
# Float division
10 / 3

3.3333333333333335

In [25]:
# Integer division

10 // 3

3

In python 2 the operators for integer and float division is reversed. So `10 / 3 == 3` and `10 // 3 == 3.333333`.

If you use Python 2 (not recommended as it will be deprecated) remember to convert your integers to floats when you want floating point precision for divisions, or use the `//` operator!

In [26]:
# convert to integer
a = 6.6
int(a)

6

In [27]:
# convert to float
b = 5
float(b)

5.0

In [29]:
10 // float(3)

3.0

We can also use comparison and logic operators:
`<, >, ==, !=, <=, >=` and statements of identity such as
`and, or, not`. The data type returned by this is 
called a _boolean_.

In [30]:
# `>` example
3 > 4

False

In [32]:
# `and` example
True and False

False

In [31]:
# `or` example
True or False

True

## Sequential types: Lists and Tuples

### Lists

**Lists** are a common data structure to hold an ordered sequence of elements.
Each element can be accessed by an index.
Note that Python indexes start with 0 instead of 1:

In [36]:
# Create a list and show the first element
numbers = [1, 2, 3]
numbers[0]

1

To add elements to the end of a list, we can use the `append` method:

In [37]:
# Append to numbers and print the full list
numbers.append(4)
numbers[3]

4

**Methods** are a way to interact with an object (a list, for example).
We can invoke a method using the dot `.` followed by the method name and a list of arguments in parentheses.
To find out what methods are available for an object, we can use the built-in `help` command:

In [40]:
# Find help for the numbers object
help(numbers)

Help on list object:

class list(object)
 |  list(iterable=(), /)
 |  
 |  Built-in mutable sequence.
 |  
 |  If no argument is given, the constructor creates a new empty list.
 |  The argument must be an iterable if specified.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(...)
 |      x.__getitem__(y) <==> x[y]
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __iadd__(self, value, /)
 |      Implement self+=value.
 |  
 |  __imul__(self, value, /)
 |      Implement self*=value.
 |  
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self)) for accurate sign

Integrated development environments (IDEs) like Jupyter often have "tab completion" or can automatically show help for you. 

In [ ]:
# Show available methods and attributes for numbers using tab
numbers.<tab>

In [ ]:
# Show help for calling functions or methods with Shift-tab
numbers.append(<Shift+tab>)

### Tuples

A tuple is similar to a list in that it's an ordered sequence of elements.
However, tuples can not be changed once created (they are "immutable").
Tuples are created by placing comma-separated values inside parentheses `()`.

In [46]:
# Define two tuples, one with ints and one with strings.
# Show the first element of one of them.
a_tuple = (1, 2, 3)
another_tuple = ("blue", "green", "red")

a_tuple[0]

1

### Challenge
1. What happens when you type `a_tuple[2] = 5` vs `a_list[1] = 5` ?
2. Type `type(a_tuple)` into python - what is the object type?


In [47]:
a_list = [1, 2, 3]
a_list[1] = 5
print(a_list)

[1, 5, 3]


In [48]:
a_tuple = (1, 2, 3)
a_tuple[1] = 5  # will fail

TypeError: 'tuple' object does not support item assignment

In [49]:
type(a_tuple)

tuple

# Pandas and DataFrames

## About Libraries

A library in Python contains a set of tools (called functions) that perform tasks on our data.
Importing a library is like getting a piece of lab equipment out of a storage locker and setting it up on the bench for use in a project.
Once a library is set up, it can be used or called to perform many tasks.

Python doesn't load all of the libraries available to it by default.
We have to add an `import` statement to our code in order to use library functions.
To import a library, we use the syntax `import libraryName`.

You only need to load a library once during your session.
You can load the library when needed or you can load all necessary libraries at the beginning of your script. 
This is good practice, especially for the readability of your code

In [56]:
# Import the builtin math library and use the sqrt function!
import math

math.sqrt(4)

2.0

In [54]:
# Demonstrate a `from` import
from math import sqrt

sqrt(4)

2.0

If we want to give the library a nickname to shorten the command, we can add `as nickNameHere`.
An example of importing the pandas library using the common nickname `pd` is below.

In [57]:
import pandas as pd

## Pandas in Python

One of the best options for working with tabular data in Python is to use the [Python Data Analysis Library](http://pandas.pydata.org/) (a.k.a. Pandas).
The Pandas library provides data structures, produces high quality plots with [matplotlib](http://matplotlib.org/) and integrates nicely with other libraries that use [NumPy](http://www.numpy.org/) (which is another Python library) arrays.

A handy **Pandas cheatsheet** can be found [here](http://pandas.pydata.org/Pandas_Cheat_Sheet.pdf).

Each time we call a function that's in a library, we use the syntax `LibraryName.FunctionName`.
Adding the library name with a `.` before the function name tells Python where to find the function.
In the example above, we have imported Pandas as `pd`.
This means we don't have to type out `pandas` each time we call a Pandas function.


## So What's a DataFrame?

A DataFrame is a 2-dimensional (labeled) data structure that stores data of different types (including characters, integers, floating point values, factors and more) in columns.

It is similar to a spreadsheet or an SQL table or the `data.frame` in R. 

A DataFrame always has an index (0-based) and works best with *tidy data*.
According to [Hadley Wickham](http://vita.had.co.nz/papers/tidy-data.html):
> Tidy datasets are easy to manipulate, model and visualize, and have a specific structure: each variable is a column, each observation is a row, and each type of observational unit is a table. 

More info on the pandas DataFrame: https://pandas.pydata.org/pandas-docs/stable/dsintro.html#dataframe


## Reading data into a pandas DataFrame

### Our Data 

For this lesson, we will be using the Portal Teaching data, a subset of the data from Ernst et al [Long-term monitoring and experimental manipulation of a Chihuahuan Desert ecosystem near Portal, Arizona, USA](http://www.esapubs.org/archive/ecol/E090/118/default.htm).

We will be using files from the [Portal Project Teaching Database](https://figshare.com/articles/Portal_Project_Teaching_Database/1314459).
This section will use the `surveys.csv` file which can be found in the `data/` directory.

We are studying the species and weight of animals caught in plots in our study
area. The dataset is stored as a `.csv` file: each row holds information for a
single animal, and the columns represent:

| Column           | Description                        |
|------------------|------------------------------------|
| record_id        | Unique id for the observation      |
| month            | month of observation               |
| day              | day of observation                 |
| year             | year of observation                |
| plot             | ID of a particular plot            |
| species          | 2-letter code                      |
| sex              | sex of animal ("M", "F")           |
| wgt              | weight of the animal in grams      |


The first few rows of our first file look like this:

```
record_id,month,day,year,plot,species,sex,wgt
1,7,16,1977,2,NA,M,
2,7,16,1977,3,NA,M,
3,7,16,1977,2,DM,F,
```

### Starting in the same spot

To help the lesson run smoothly, let's ensure everyone is in the same directory.
This should help us avoid path and file name issues. 

At this time please navigate to the workshop directory (ie, the directory with the notebooks in them). 
If you working in IPython Notebook be sure that you start your notebook in the workshop directory.


A quick aside, there are Python libraries like [`os`](https://docs.python.org/3/library/os.html) that can work with our
directory structure, however, that is not our focus today.


If you **need to change your directory** `import os` and use `os.chdir`

Or you can use `%%bash` cell magic to access the command line, e.g.

```
%%bash

cd folder_name
```

In [ ]:
# if you need to change your directory
import os
os.getcwd()

In [ ]:
os.listdir("./")

In [ ]:
import pandas as pd

# check your version, we need v0.19 or higher
pd.__version__


We will begin by locating and reading our survey data which are in CSV format.
We can use Pandas' `read_csv` function to pull the file directly into a
[DataFrame](http://pandas.pydata.org/pandas-docs/stable/dsintro.html#dataframe)

In [ ]:
# note that pd.read_csv is used because we imported pandas as pd



We can see that there were 33,549 rows parsed. Each row has 9
columns. The first column is the index of the DataFrame. The index is used to
identify the position of the data, but it is not an actual column of the DataFrame. 
It looks like  the `read_csv` function in Pandas  read our file properly. However, 
we haven't saved any data to memory so we can work with it.We need to assign the 
DataFrame to a variable. Remember that a variable is a name for a value, such as `x`, 
or  `data`. We can create a new  object with a variable name by assigning a value to it using `=`.

Let's call the imported survey data `surveys_df`:



Notice when you assign the imported DataFrame to a variable, Python does not
produce any output on the screen. We can print the value of the `surveys_df`
object by typing its name into the Python command prompt.


## Manipulating Our Species Survey Data

Now we can start manipulating our data. First, let's check the data type of the
data stored in `surveys_df` using the `type` method. The `type` method and
`__class__` attribute tell us that `surveys_df` is `<class 'pandas.core.frame.DataFrame'>`.

In [ ]:
type(surveys_df)
# or surveys_df.__class__

We can also enter `surveys_df.dtypes` at our prompt to view the data type for each
column in our DataFrame. `int64` represents numeric integer values - `int64` cells
can not store decimals. `object` represents strings (letters and numbers). `float64`
represents numbers with decimals.

In [ ]:
surveys_df.dtypes

Pandas and base Python use slightly different names for data types. More on this
is in the table below:

| Pandas Type | Native Python Type | Description |
|-------------|--------------------|-------------|
| object | string | The most general dtype. Will be assigned to your column if column has mixed types (numbers and strings). |
| int64  | int | Numeric characters. 64 refers to the memory allocated to hold this character. |
| float64 | float | Numeric characters with decimals. If a column contains numbers and NaNs(see below), pandas will default to float64, in case your missing value has a decimal. |
| datetime64, timedelta[ns] | N/A (but see the [datetime] module in Python's standard library) | Values meant to hold time data. Look into these for time series experiments. |

[datetime]: http://doc.python.org/2/library/datetime.html


---

### Challenges

Try out the methods below to see what they return.

1. `surveys_df.columns`.
2. `surveys_df.head()`. Also, what does `surveys_df.head(15)` do?
3. `surveys_df.tail()`.
4. `surveys_df.shape`. Take note of the output of the shape method. What format does it return the shape of the DataFrame in?

HINT: [More on tuples, here](https://docs.python.org/3/tutorial/datastructures.html#tuples-and-sequences).

## Exploring DataFrames in Python

There are multiple methods that can be used to access and summarise the data
stored in DataFrames. Let's try out a few. Note that we call the method by using
the object name followed by . and the method name. So `surveys_df.columns` provides an index
of all of the column names in our DataFrame.

### Selecting Data Using Labels (Column Headings)


In pandas you can use several ways to **select a specific column**:
- square brackets `[]` 
- a `.` and the column name

For example, we can select all of data from a column named `species` from the `surveys_df`
DataFrame by name:

```python
surveys_df['species']
# this syntax, calling the column as an attribute, gives you the same output
surveys_df.species
```

Using double square brackets `[[]]` we can pass a list of column names too by listing the names we want:

We can also create an new object that contains the data as follows:

```python
# create an object named surveys_species that only contains the `species_id` column
surveys_species = surveys_df['species']
```

**NOTE:** If a column name is not contained in the DataFrame, an exception
(error) will be raised.

```python
surveys_df['speciess']
```

---

Converting between different data types

In [ ]:
# convert the record_id field from an integer to a float


What happens if we try to convert weight values to integers?

Notice that this throws a value error: `ValueError: Cannot convert NA to
integer`. If we look at the `weight` column in the surveys data we notice that
there are NaN (Not A Number) values. *NaN* values are undefined
values that cannot be represented mathematically. Pandas, for example, will read
an empty cell in a CSV or Excel sheet as a NaN. NaNs have some desirable
properties: if we were to average the `weight` column without replacing our NaNs,
Python would know to skip over those cells.


_Note: older pandas version do not know how to handle NaN, please update to v0.19 or higher_

Check your pandas version using `pd.__version__`, if you need to update open a bash shell
and type ```conda update pandas```.

---

## Missing Data Values - NaN

Dealing with missing data values is always a challenge. It's sometimes hard to
know why values are missing - was it because of a data entry error? Or data that
someone was unable to collect? Should the value be 0? We need to know how
missing values are represented in the dataset in order to make good decisions.
If we're lucky, we have some metadata that will tell us more about how null
values were handled.

For instance, in some disciplines, like Remote Sensing, missing data values are
often defined as -9999. Having a bunch of -9999 values in your data could really
alter numeric calculations. Often in spreadsheets, cells are left empty where no
data are available. Pandas will, by default, replace those missing values with
NaN. However it is good practice to get in the habit of intentionally marking
cells that have no data, with a no data value! That way there are no questions
in the future when you (or someone else) explores your data.

### Where Are the NaN's?

Let's explore the NaN values in our data a bit further. 
First, let's figure out **how many rows contain NaN values for weight**. 
We can do this by identifying how many rows have a NULL value (`.isnull`) or by counting the number of rows that have a meaningful value (e.g., wgt>0):

We can replace all NaN values with zeroes using the `.fillna()` method (after
making a copy of the data so we don't lose our work).

However, `NaN` and 0 yield different analysis results. The mean value when `NaN`
values are replaced with 0 is different from when `NaN` values are simply thrown
out or ignored.

In [ ]:
#replace NaN with 0


In [ ]:
#check mean, how does it differ from before?


We can fill `NaN` values with any value that we chose.
The code below fills all `NaN` values with a mean for all weight values.

```python
 df1['wgt'] = surveys_df['wgt'].fillna(surveys_df['wgt'].mean())
```

We could also chose to create a subset of our data, only keeping rows that do not contain `NaN` values, using `.dropna()` method.

**The point is to make conscious decisions about how to manage missing data.**

This is where we think about how our data will be used and how these values will
impact the scientific conclusions made from the data.

Python gives us all of the tools that we need to account for these issues. We
just need to be cautious about how the decisions that we make impact scientific
results.

## Calculating summary statistics for a Pandas DataFrame

We've read our data into Python. Next, let's perform some quick summary
statistics to learn more about the data that we're working with. We might want
to know how many animals were collected in each plot, or how many of each
species were caught. We can perform summary stats quickly using groups. But
first we need to figure out what we want to group by.

---

Let's find out how many unique plot IDs and species we have in our data:

In [ ]:
# Reminder of the column names


In [ ]:
# Create a list of unique plot ID's  and species found in the surveys data


In [ ]:
# Check the length of the list
print("There are: " + str(len(plot_names)) + " unique plots in the data")
print("There are: " + str(len(species)) + " unique species in the data")

In [ ]:
# Single line solution
print("There are: " + str(surveys_df['plot'].nunique()) + " unique plots in the data")
print("There are: " + str(surveys_df['species'].nunique()) + " unique species in the data")

---

The Pandas function `describe` will return descriptive stats including: mean,
median, max, min, std and count for a particular column in the data. Pandas'
`describe` function will only return summary values for columns containing
numeric data.
We can calculate basic statistics for all records in a single column using the
syntax below:

We can also extract one specific metric if we wish:

```python
surveys_df['wgt'].min()
surveys_df['wgt'].max()
surveys_df['wgt'].mean()
surveys_df['wgt'].std()
surveys_df['wgt'].count()
```


### Basic Math Functions

If we wanted to, we could perform math on an entire column of our data. For
example let's multiply all weight values by 2. A more practical use of this might
be to normalize the data according to a mean, area, or some other value
calculated from our data.

In [ ]:
# multiply all weight values by 2


### Groups in Pandas

We often want to calculate summary statistics grouped by subsets or attributes
within fields of our data, for example we might want to know what the summary stats look like split by sex.
We can use Pandas' `.groupby` method, which creates a groupby DataFrame on which we can perform other pandas methods.


In [ ]:
# grouping the df by sex

# summary statistics for this new df



In [ ]:
# provide the min for each numeric column by sex



The `groupby` command is powerful in that it allows us to quickly generate
summary stats, not just for one group but several.

For example, we might want to calculate the average
weight of all individuals per plot:

```python
surveys_df.groupby('plot')['wgt'].mean()
```

In [ ]:
# calculate average weight of individuals in each plot



Or, we might want to know how many males and females we have for each species:

```python
surveys_df.groupby(['species','sex'])['record_id'].count()
```

In [ ]:
# count the number of each sex per species



Another quick way to get the count of each unique value in a column is the `value_counts()` method:

### Challenge


1. What happens when you group by two columns using the following syntax and
    then grab mean values:
	- `sorted_data = surveys_df.groupby(['plot','species'])`
	- `sorted_data.mean()`
2. Get summary statistics on the weight for grouped by the plot, species and sex. HINT: you can use the
   following syntax to only create summary statistics for one column in your data
   `name_of_dataframe['name_of_column'].describe()`


## Quick & Easy Plotting Data Using Pandas

We can plot our summary stats using Pandas, too.

In [ ]:
# make sure figures appear inline in Ipython Notebook
%matplotlib inline

# plot year vs wgt



In [ ]:
# create a quick bar chart



In [ ]:
# We can also look at how many animals were captured in each plot:

# let's plot that too



### Challenge Activities

1. Create a scatter plot of average weight across all species per plot. x-axis = plot, y-axis = wgt
2. Create the same plot, but with average weight for each sex per plot. Hint, you will need to `unstack` when plotting. x-axis = plot, y-axis = wgt, different lines for each sex.
3. Create a trend plot of the average weight per plot over time. x-axis = year, y-axis = wgt, different lines for each plot.

In [ ]:
# group by plot and calculate mean wgt


# let's plot, you should have x-axis = plot, y-axis = wgt



In [ ]:
# group by plot and sex, then calculate mean wgt


# let's plot, you should have x-axis = plot, y-axis = wgt, different lines for sex
# you need to use the .unstack() method before the .plot() for this to work



In [ ]:
# group by year and plot, then calculate mean wgt


# let's plot, you should have x-axis = year, y-axis = wgt, different lines for plot
# you need to use the .unstack() method before the .plot() for this to work



# Indexing & Slicing in Python

We often want to work with subsets of a **DataFrame** object. There are
different ways to accomplish this including: using labels (ie, column headings - as used previously),
numeric ranges or specific x,y index locations.

## Extracting Range based Subsets: Slicing

**REMINDER**: Python Uses 0-based Indexing

Let's remind ourselves that Python uses 0-based
indexing. This means that the first element in an object is located at position
0. This is different from other tools like R and Matlab that index elements
within objects starting at 1.



![indexing diagram](https://datacarpentry.org/python-ecology-lesson/fig/slicing-indexing.png)

![slicing diagram](https://datacarpentry.org/python-ecology-lesson/fig/slicing-slicing.png)

## Challenges

```python
# Create a list of numbers:
a = [1,2,3,4,5]
```

1. What value does the code below return?
        a[0]
2. How about this:
        a[5]
3. Or this?
        a[len(a)]
4. In the example above, calling `a[5]` returns an error. Why is that?

## Slicing using the `[]` operator

Slicing using the `[]` operator selects a set of rows and/or columns from a
DataFrame. To slice out a set of rows, you use the following syntax:
`data[start:stop]`. When slicing in pandas the start bound is included in the
output. The stop bound is one step BEYOND the row you want to select. So if you
want to select rows 0, 1 and 2 your code would look like this:


```python
# select rows 0,1,2 (but not 3)
surveys_df[0:3]
```

The stop bound in Python is different from what you might be used to in
languages like Matlab and R.

```python
# select the first, second and third rows from the surveys variable
surveys_df[0:3]
# select the first 5 rows (rows 0,1,2,3,4)
surveys_df[:5]
# select the last element in the list
surveys_df[-1:]
```

To access columns we will need to call them by their labels, if you want more than one column you need to supply a list of column names.

```python
# selecting the plot column
surveys_df['plot']

# selecting plot and weight
surveys_df[['plot','wgt']]
```


In [ ]:
surveys_df[-1:]

In [ ]:
surveys_df['plot'][[0,1,2]]

In [ ]:
surveys_df['plot'][0:3]

In [ ]:
surveys_df[['plot','wgt']][0:3]

We can also reassign values within subsets of our DataFrame. But before we do that, let's make a 
copy of our DataFrame so as not to modify our original imported data. 

```python
# copy the surveys dataframe so we don't modify the original DataFrame
surveys_copy = surveys_df

# set the first three rows of data in the DataFrame to 0
surveys_copy[0:3] = 0
```

Next, try the following code: 

```python
surveys_copy.head()
surveys_df.head()
```
What is the difference between the two data frames?

## Referencing Objects vs Copying Objects in Python

We might have thought that we were creating a fresh copy of the `surveys_df` objects when we  used the code `surveys_copy = surveys_df`.
However the statement `y = x` doesn’t create a copy of our DataFrame.
It creates a new variable `y` that refers to the **same** object `x` refers to.
This means that there is only one object (the DataFrame), and both `x` and `y` refer to it.
So when we assign the first 3 columns the value of `0` using the `surveys_copy` DataFrame, the `surveys_df` DataFrame is modified too.

To create a fresh copy of the `surveys_df` DataFrame we use the syntax `y = x.copy()`.
But before we have to read the `surveys_df` again because the current version contains the unintentional changes made to the first 3 columns.

```python
surveys_df = pd.read_csv("data/surveys.csv")
surveys_copy = surveys_df.copy()
```

## Slicing and subsetting: label vs integer-based indexing

We can select specific loctions or ranges of our data in both the row and column directions
using either label or integer-based indexing.

- `loc`: indexing via *labels* (which can be numbers)
- `iloc`: indexing via *integers*


![loc_iloc_subsetting](http://104.236.88.249/wp-content/uploads/2016/10/Pandas-selections-and-indexing.png)


![dataframe_indexing](https://vrzkj25a871bpq7t1ugcgmn9-wpengine.netdna-ssl.com/wp-content/uploads/2019/01/pandas-dataframe-has-indexes.png)


To select an index subset of rows AND columns from our DataFrame, we can use the `iloc` method.
For example, we can select RA, Dec and number of votes (columns 2, 3 and 4 if we start counting at 1), by slicing our index like this:

```python
surveys_df.iloc[0:3, 1:4]
```

**Note**: the order of selection is ROW followed by COLUMN

In [ ]:
# if you haven't done so yet read the csv back in (we had unwanted changes when 
# we 'copied' the data frame earlier)
surveys_df = pd.read_csv("data/surveys.csv")

#try .iloc
surveys_df.iloc[0:3 , 1:4]

Notice that we asked for a slice from `0:3`. This yielded 3 rows of data. When you
ask for `0:3`, you are actually telling python to start at index 0 and select rows
`0`, `1`, and `2` (**up to but not including 3**).


---

Next let's explore subsetting our data using labels.
**Note** When "slicing" labels, the start bound and the stop bound are **included**. 

In [ ]:
# select all columns for rows of index values 0 and 10
surveys_df.loc[[0, 10], :]

In [ ]:
# what does this do?
surveys_df.loc[0, ['species', 'plot', 'wgt']]


In [ ]:
# What happens when you type the code below?
surveys_df.loc[[0, 10, 35549], :]


In [ ]:
# Try using iloc instead?
surveys_df.iloc[[0, 10, 35549], :]


NOTE: If using `iloc` labels must be found in the DataFrame or you will get a `KeyError`.
Using `loc` (at least for now) you'll get `NaN` entries returned so be careful!


## Challenge Activities

1. What happens when you type:
	- `surveys_df[0:3]`
	- `surveys_df[:5]`
	- `surveys_df[-1:]`

2. What happens when you call:
    - `dat.iloc[0:4, 1:4]`
    - `dat.loc[0:4, 1:4]`
    - How are the two commands different?

## Subsetting using masks

A mask can be useful to locate where a particular subset of values exist or don't exist - for example, `NaN`, or "Not a Number" values.

To understand masks we also need to understand `BOOLEAN` objects in python.

Boolean values include `True` or `False`.
So for example:

```python
# set x to 5
x = 5

# what does the code below return?
x > 5

# how about this?
x == 5
```

When we ask python what the value of `x > 5` is, we get `False`.
This is because `x` is not greater than 5 it is equal to 5.
To create a boolean mask, you first create the `True` / `False` criteria (e.g. `values > 5`).
Python will then assess each value in the object to determine whether the value meets the criteria (`True`) or not (`False`).
Python creates an output object that is the same shape as the original object, but with a `True` or `False` value for each index location.

You can use the syntax below when querying data from a DataFrame.
Experiment with selecting various subsets of our data.

* Equals: `==`
* Not equals: `!=`
* Greater than, less than: `>` or `<`
* Greater than or equal to `>=`
* Less than or equal to `<=`

Let's try this out. 

Next, let's identify all locations in the survey data that have null (missing or `NaN`) data values. We can use the `isnull` method to do this.
Each cell with a null value will be assigned a value of  `True` in the new boolean object.

To select the rows where there are null values,  we can use 
the mask as an index to subset our data as follows:

```python
# To select just the rows with NaN values, we can use the .any method
surveys_df[pd.isnull(surveys_df).any(axis=1)]
```

Note that there are many null or `NaN` values in the `wgt` column of our DataFrame.
We will explore different ways of dealing with these in Lesson 03.

We can run `isnull` on a particular column too.
What does the code below do?

```python
# what does this do?
empty_weights = surveys_df[pd.isnull(surveys_df).any(axis=1)]['wgt']
```

Let's take a minute to look at the statement above. 

We are using the Boolean object as an index. 
We are asking python to select rows that have a `NaN` value for weight.

---

We can also select a subset of our data using criteria. For example, we can
select all rows that have a year value of 2002.

In [ ]:
#Or we can select all rows that do not contain the year 2002.



In [ ]:
#We can define sets of criteria too:



### Challenge Activities

1. Select a subset of rows in the `surveys_df` DataFrame that contain data from
   the year 1999 and that contain weight values less than or equal to 8. How
   many columns did you end up with? What did your neighbor get?
2. You can use the `isin` command in python to query a DataFrame based upon a
   list of values as follows:
   `surveys_df[surveys_df['species'].isin([listGoesHere])]`. Use the `isin` function
   to find all plots that contain particular species in
   the surveys DataFrame. How many records contain these values?
3. Experiment with other queries. Create a query that finds all rows with a weight value > or equal to 0.
4. The `~` symbol in Python can be used to return the OPPOSITE of the selection that you specify in python. 

It is equivalent to **is not in**.
Write a query that selects all rows that are NOT equal to 'M' or 'F' in the surveys data.

In [ ]:
# what about our NaN?

## Concatenating

We can use the `concat` function in Pandas to append either columns or rows from one DataFrame to another.
Let's grab two subsets of our data to see how this works.

In [ ]:
# read in first 10 lines of surveys table



In [ ]:
# grab the last 10 rows (minus the last one)



In [ ]:
#reset the index values to the second dataframe appends properly
# drop=True option avoids adding new index column with old index values



When we concatenate DataFrames, we need to specify the axis.

`axis=0` tells Pandas to stack the second DataFrame under the first one.
It will automatically detect whether the column names are the same and will stack accordingly.

`axis=1` will stack the columns in the second DataFrame to the RIGHT of the first DataFrame.
To stack the data vertically, we need to make sure we have the same columns and associated column format in both datasets.
When we stack horizonally, we want to make sure what we are doing makes sense (ie the data are related in some way).

In [ ]:
# stack the DataFrames on top of each other



In [ ]:
# place the DataFrames side by side



### Challenge - Row Index Values and Concat
Have a look at the `vertical_stack` dataframe, do you notice anything unusual?

The row indexes for the two data frames `survey_sub` and `survey_sub_last10` have been repeated.
We can reindex the new dataframe using the `reset_index()` method.

## Writing Out Data to CSV

We can use the `to_csv` command to do export a DataFrame in CSV format.
Note that the code below will by default save the data into the current working directory.
We can save it to a different folder by adding the foldername and a slash to the file `vertical_stack.to_csv('foldername/out.csv')`.

```python
# Write DataFrame to CSV 
vertical_stack.to_csv('out.csv')
```

Check out your working directory to make sure the CSV wrote out properly, and that you can open it!
If you want, try to bring it back into python to make sure it imports properly.

```python	
# for kicks read our output back into python and make sure all looks good
new_output = pd.read_csv('out.csv', keep_default_na=False, na_values=[""])
```

# Automating data processing using For Loops

So far, we've used Python and the pandas library to explore and manipulate individual datasets by hand, much like we would do in a spreadsheet.

The beauty of using a programming language like Python comes from the ability to automate data processing through the use of loops and functions.


## For loops

Loops allow us to repeat a workflow (or series of actions) a given number of times or while some condition is true.
We would use a loop to automatically process data that's stored in multiple files E.G. daily values with one file per year.

Loops lighten our work load by performing repeated tasks without our direct involvement and make it less likely that we'll introduce errors by making mistakes while processing each file by hand.

Let's write a simple for loop that simulates what a kid might see during a visit to the zoo:

In [ ]:
animals = ['lion', 'tiger', 'crocodile', 'vulture', 'hippo']
print(animals)

In [ ]:
for creature in animals:
    print(creature)

The line defining the loop must start with `for` and end with a colon, and the body of the loop must be indented.

In this example, `creature` is the loop variable that takes the value of the next entry in `animals` every time the loop goes around.

We can call the loop variable anything we like.
After the loop finishes, the loop variable will still exist and will have the value of the last entry in the collection:

In [ ]:
for creature in animals:
    pass

In [ ]:
creature

We are not asking python to print the value of the loop variable anymore, but the for loop still runs and the value of `creature` changes on each pass through the loop.
The statement `pass` in the body of the loop just means "do nothing".

---

The file we've been using so far (`surveys.csv`) contains 25 years of data and is
very large. We would like to separate the data for each year into a separate
file.

Let's start by making a new directory called `yearly_files` inside the folder `data` to store all of these files using the module `os`:

The command `os.mkdir` is equivalent to `mkdir` in the shell. Just so we are
sure, we can check that the new directory was created within the `data` folder:

The command `os.listdir` is equivalent to `ls` in the shell.

Previously, we saw how to use the library pandas to load the species
data into memory as a DataFrame, how to select a subset of the data using some
criteria, and how to write the DataFrame into a csv file. Let's write a script
that performs those three steps in sequence for the year 2002:

```python
import pandas as pd

# Load the data into a DataFrame
surveys_df = pd.read_csv('data/surveys.csv')

# Select only data for 2002
surveys2002 = surveys_df[surveys_df.year == 2002]

# Write the new DataFrame to a csv file
surveys2002.to_csv('data/yearly_files/surveys2002.csv')
```

To create yearly data files, we could repeat the last two commands over and over, once for each year of data.
Repeating code is neither elegant nor practical, and is very likely to introduce errors into your code.

We want to turn what we've just written into a loop that repeats the last two commands for every year in the dataset.

Let's start by writing a loop that simply prints the names of the files we want to create - the dataset we are using covers 1977 through 2002, and we'll create a separate file for each of those years.
Listing the filenames is a good way to confirm that the loop is behaving as we expect.

We have seen that we can loop over a list of items, so we need a list of years to loop over.

We can get the *unique* years in our DataFrame with:

Putting this into our for loop we get

Notice that we use single quotes to add text strings. The variable is not
surrounded by quotes. This code produces the string
`data/yearly_files/surveys_year2002.csv` which contains the path to the new filename
AND the file name itself.

We can now add the rest of the steps we need to create separate text files.
Once finished look inside the `yearly_files` directory and check a couple of the files you
just created to confirm that everything worked as expected.

In [ ]:
for year in surveys_df['year'].unique():
    #creating filename
    filename = 'data/yearly_files/surveys_year' + str(year) + '.csv'
    # extracting data of a specific year
    surveys_year = surveys_df[surveys_df['year']==year]
    # writing to file
    surveys_year.to_csv(filename)

In [ ]:
os.listdir('data/yearly_files/')

### Challenge

1. What happens if there is no data for a year in the sequence (for example, imagine we had used 1976 as the start year in `range`)?

2. Let's say you only want to look at data from a given multiple of years. How would you modify your loop in order to generate a data file for only every 5th year, starting from 1977? Hint: you will need to use `range` to specify the list of numbers.

``` python
range(start, end, steps)
```

3. Instead of splitting out the data by years, a colleague wants to do analyses each species separately. How would you write a unique csv file for each species?

In [ ]:
surveys_df[(surveys_df['year']== 1976)]

In [ ]:
for year in range(surveys_df['year'].min(), surveys_df['year'].max(), 5):
    #creating filename
    filename = 'data/yearly_files/surveys' + str(year) + '.csv'

    # extracting data of a specific year
    surveys_year = surveys_df[surveys_df['year'] == year]

    # writing to file
    surveys_year.to_csv(filename)

In [ ]:
# Make a folder to store species tables!

In [ ]:
for species in surveys_df.species.dropna().unique():
    #creating filename
    filename = 'data/species/surveys_species_' + species + '.csv'

    # extracting data of a specific year
    surveys_species = surveys_df[surveys_df.species == species]

    # writing to file
    surveys_species.to_csv(filename)

## Building reusable and modular code with functions

Suppose that separating large data files into individual yearly files is a task that we frequently have to perform.
We could write a **for loop** like the one above every time we needed to do it but that would be time consuming and error prone.
A more elegant solution would be to create a reusable tool that performs this task with minimum input from the user.
To do this, we are going to turn the code we've already written into a function.

Functions are reusable, self-contained pieces of code that are called with a single command.
They can be designed to accept arguments as input and return values, but they don't need to do either.
Variables declared inside functions only exist while the function is running and if a variable within the function (a local variable) has the same name as a variable somewhere else in the code, the local variable hides but doesn't overwrite the other.

Every method used in Python (for example, `print`) is a function, and the libraries we import (say, `pandas`) are a collection of functions.

We will only use functions that are housed within the same code that uses them, but it's possible [easy to write functions that can be used by different programs](https://docs.python.org/3/tutorial/modules.html).


Functions are declared following this general structure:

```python
def this_is_the_function_name(input_argument1, input_argument2):
    # The body of the function is indented

    # This function prints the two arguments to screen
    print(
        'The function arguments are:',
        input_argument1,
        input_argument2,
        '(this is done inside the function!)'
    )

    # And returns their product
    return input_argument1 * input_argument2
```

The function declaration starts with the word `def`, followed by the function
name and any arguments in parenthesis, and ends in a colon. The body of the
function is indented just like loops are. If the function returns something when
it is called, it includes a return statement at the end.

In [ ]:
# let's define this function
def this_is_the_function_name(input_argument1, input_argument2):
    """
    input_argument1: a float or numerical type
    """

    # The body of the function is indented
    # This function prints the two arguments to screen
    print(
        'The function arguments are:',
        input_argument1,
        input_argument2,
        '(this is done inside the function!)'
    )

    # And returns their product
    return input_argument1 * input_argument2

In [ ]:
# and now let's call the function:
result = this_is_the_function_name(4,4)
result

### Challenge:

2. Try calling the function by giving it the wrong number of arguments (not 2)
   or not assigning the function call to a variable (no `product_of_inputs =`)
3. Declare a variable inside the function and test to see where it exists (Hint:
   can you print it from outside the function?)
4. Explore what happens when a variable both inside and outside the function
   have the same name. What happens to the global variable when you change the
   value of the local variable?

In [ ]:
this_is_the_function_name(1)

In [ ]:
def this_other_function(in1, in2, in3=74646):
    new_variable = 3
    print(new_variable, in1, in2, in3)
    return

In [ ]:
this_other_function(1, 2)

In [ ]:
new_variable = 5
print(new_variable)

this_other_function(1, 2)
print(new_variable)

We can now turn our code for saving yearly data files into a function. There are
many different "chunks" of this code that we can turn into functions, and we can
even create functions that call other functions inside them. Let's first write a
function that separates data for just one year and saves that data to a file:

```python
def one_year_csv_writer(this_year, all_data):
    """
    Writes a csv file for data from a given year.

    this_year --- year for which data is extracted
    all_data --- DataFrame with multi-year data
    """

    # Select data for the year
    surveys_year = all_data[all_data.year == this_year]

    # Write the new DataFrame to a csv file
    filename = 'data/yearly_files/function_surveys_year' + str(this_year) + '.csv'
    surveys_year.to_csv(filename)
```

In [ ]:
def one_year_csv_writer(this_year, all_data):
    """
    Writes a csv file for data from a given year.

    this_year --- year for which data is extracted
    all_data --- DataFrame with multi-year data
    """

    # Select data for the year
    surveys_year = all_data[all_data.year == this_year]

    # Write the new DataFrame to a csv file
    filename = 'data/yearly_files/function_surveys_year' + str(this_year) + '.csv'
    surveys_year.to_csv(filename)

In [ ]:
for year in surveys_df['year'].unique():
    one_year_csv_writer(year, surveys_df)

The text between the two sets of triple double quotes is called a docstring and contains the documentation for the function.
It does nothing when the function is running and is therefore not necessary, but it is good practice to include docstrings as a reminder of what the code does.

Docstrings in functions also become part of their 'official' documentation:

In [ ]:
one_year_csv_writer?

In [ ]:
help(one_year_csv_writer)

In [ ]:
one_year_csv_writer(2002, surveys_df)

We changed the root of the name of the csv file so we can distinguish it from the one we wrote before.
Check the `yearly_files` directory for the file.
Did it do what you expect?

What we really want to do, though, is create files for multiple years without having to request them one by one.
Let's write another function that replaces the entire For loop by simply looping through a sequence of years and repeatedly calling the function we just wrote, `one_year_csv_writer`:


```python
def yearly_data_csv_writer(start_year, end_year, all_data):
    """
    Writes separate csv files for each year of data.

    start_year --- the first year of data we want
    end_year --- the last year of data we want
    all_data --- DataFrame with multi-year data
    """

    # "end_year" is the last year of data we want to pull, so we loop to end_year+1
    for year in range(start_year, end_year+1):
        one_year_csv_writer(year, all_data)
```

In [ ]:
def yearly_data_csv_writer(start_year, end_year, all_data):
    """
    Writes separate csv files for each year of data.

    start_year --- the first year of data we want
    end_year --- the last year of data we want
    all_data --- DataFrame with multi-year data
    """

    # "end_year" is the last year of data we want to pull, so we loop to end_year+1
    for year in range(start_year, end_year+1):
        one_year_csv_writer(year, all_data)

Because people will naturally expect that the end year for the files is the last
year with data, the for loop inside the function ends at `end_year + 1`. 
This is because when we specify `range()` the last number is not included, try it for yourself.

In [ ]:
range(5)

By writing the entire loop into a function, we've made a reusable tool for whenever
we need to break a large data file into yearly files. Because we can specify the
first and last year for which we want files, we can even use this function to
create files for a subset of the years available. This is how we call this
function:

In [ ]:
yearly_data_csv_writer(1980, 1990, surveys_df)

**BEWARE!** If you are using IPython Notebooks and you modify a function, you MUST
re-run that cell in order for the changed function to be available to the rest
of the code. Nothing will visibly happen when you do this, though, because
simply defining a function without *calling* it doesn't produce an output. Any
cells that use the now-changed functions will also have to be re-run for their
output to change.

### Challenge:

1. **Add two arguments** to the functions we wrote that take the *path* of the
   directory where the files will be written and the *root* of the file name.
   Create a new set of files with a different name in a different directory.
2. Make the functions **return a list** of the files they have written. There are
   many ways you can do this (and you should try them all!): either of the
   functions can print to screen, either can use a return statement to give back
   numbers or strings to their function call, or you can use some combination of
   the two. You could also try using the `os` library to list the contents of
   directories.

In [ ]:
# adding two arguments
def one_year_csv_writer(all_data, directory, file_root, this_year=1977):
    """
    Writes a csv file for data from a given year in the speciefied directory using specified root name

    this_year --- year for which data is extracted --- default: 1977
    all_data --- DataFrame with multi-year data
    directory --- directory in which the data is to be saved, include the /
    file_root --- prefix for the filename [prefix_year.csv], include any _ wanted
    """

    # Select data for the year
    surveys_year = all_data[all_data.year == this_year]

    # Write the new DataFrame to a csv file
    filename =  directory + file_root + str(this_year) + '.csv'
    surveys_year.to_csv(filename)


def yearly_data_csv_writer(all_data, directory, file_root, start_year=1977, end_year=2002):
    """
    Writes separate csv files for each year of data.

    start_year --- the first year of data we want --- default: 1977
    end_year --- the last year of data we want --- default: 2002
    all_data --- DataFrame with multi-year data
    directory --- directory in which the data is to be saved, include the /
    file_root --- prefix for the filename [prefix_year.csv], include any _ wanted
    """

    # "end_year" is the last year of data we want to pull, so we loop to end_year+1
    for year in range(start_year, end_year+1):
        one_year_csv_writer(year, all_data, directory, file_root)

In [ ]:
# adding the list of filenames
def one_year_csv_writer(all_data, directory, file_root, this_year=1977):
    """
    Writes a csv file for data from a given year in the speciefied directory using specified root name

    this_year --- year for which data is extracted --- default: 1977
    all_data --- DataFrame with multi-year data
    directory --- directory in which the data is to be saved, include the /
    file_root --- prefix for the filename [prefix_year.csv], include any _ wanted
    """

    # Select data for the year
    surveys_year = all_data[all_data.year == this_year]

    # Write the new DataFrame to a csv file
    filename =  directory + file_root + str(this_year) + '.csv'
    surveys_year.to_csv(filename)
    return(filename)


def yearly_data_csv_writer(all_data, directory, file_root, start_year=1977, end_year=2002):
    """
    Writes separate csv files for each year of data.

    start_year --- the first year of data we want --- default: 1977
    end_year --- the last year of data we want --- default: 2002
    all_data --- DataFrame with multi-year data
    directory --- directory in which the data is to be saved, include the /
    file_root --- prefix for the filename [prefix_year.csv], include any _ wanted
    """
    fname = []
    # "end_year" is the last year of data we want to pull, so we loop to end_year+1
    for year in range(start_year, end_year+1):
        fname.append(one_year_csv_writer(year, all_data, directory, file_root))
    return(fname)

---

But what if our dataset doesn't start in 1977 and end in 2002? We can modify the
function so that it looks for the start and end years in the dataset if those
dates are not provided:

```python
    def yearly_data_arg_test(all_data, start_year=None, end_year=None):
        """
        Modified from yearly_data_csv_writer to test default argument values!

        start_year --- the first year of data we want --- default: None - check all_data
        end_year --- the last year of data we want --- default: None - check all_data
        all_data --- DataFrame with multi-year data
        """

        if not start_year:
            start_year = min(all_data.year)
        if not end_year:
            end_year = max(all_data.year)

        return start_year, end_year
```

In [ ]:
# define function
def yearly_data_arg_test(all_data, start_year=None, end_year=None):
    """
    Modified from yearly_data_csv_writer to test default argument values!

    start_year --- the first year of data we want --- default: None - check all_data
    end_year --- the last year of data we want --- default: None - check all_data
    all_data --- DataFrame with multi-year data
    """

    if not start_year:
        start_year = min(all_data.year)
    if not end_year:
        end_year = max(all_data.year)

    return start_year, end_year

In [ ]:
# test function
yearly_data_arg_test(surveys_df)

The default values of the `start_year` and `end_year` arguments in the function `yearly_data_arg_test` are now `None`.
This is a build-it constant in Python
that indicates the absence of a value - essentially, that the variable exists in
the namespace of the function (the directory of variable names) but that it
doesn't correspond to any existing object.

The body of the test function now has two conditional 'loops' (**if statement**) that
check the values of `start_year` and `end_year`. If statements execute the body of
the 'loop' when some condition is met. 

`if statements` work like the boolean logic we saw earlier when we created masks to select our data.
As a function they commonly look something like this:

```python
    a = 5

    if a<0: # meets first condition?

        # if a IS less than zero
        print('a is a negative number')

    elif a>0: # did not meet first condition. meets second condition?

        # if a ISN'T less than zero and IS more than zero
        print('a is a positive number')

    else: # met neither condition

        # if a ISN'T less than zero and ISN'T more than zero
        print('a must be zero!')
```

In [ ]:
a = 5

if a < 0: # meets first condition?

    # if a IS less than zero
    print('a is a negative number')

elif a > 0: # did not meet first condition. meets second condition?

    # if a ISN'T less than zero and IS more than zero
    print('a is a positive number')

else: # met neither condition

    # if a ISN'T less than zero and ISN'T more than zero
    print('a must be zero!')

Change the value of `a` to see how this function works. The statement `elif`
means "else if", and all of the conditional statements must end in a colon.

The if statements in the function `yearly_data_arg_test` check whether there is an
object associated with the variable names `start_year` and `end_year`. If those
variables are `None`, the if statements return the boolean `True` and execute whatever
is in their body. On the other hand, if the variable names are associated with
some value (they got a number in the function call), the if statements return `False`
and do not execute. The opposite conditional statements, which would return
`True` if the variables were associated with objects (if they had received value
in the function call), would be `if start_year` and `if end_year`.

### Challenge:

1. Rewrite the `one_year_csv_writer` and `yearly_data_csv_writer` functions to use `none` as default for the years.


3. The code below checks to see whether a directory exists and creates one if it
doesn't. Add some code to your function that writes out the CSV files, to check
for a directory to write to.

```Python
	if 'dir_name_here' in os.listdir('.'):
	    print('Processed directory exists')
	else:
	    os.mkdir('dir_name_here')
	    print('Processed directory created')
```

2. Modify the functions so that they don't create yearly files if there is no
data for a given year and display an alert to the user (Hint: use conditional
statements to do this.)